### PR TITLE
PLAT-338: StateReport SM

### PR DESCRIPTION
**Pull request by sexyhamster**
**- What I did**
- new sm for SMObject that goes beyong one pulse
- SMObject creates UnboundSharedData with key {reference,pulse.Number}

**- How to verify it**
run tests

### CODE REVIEW

File: ledger-core/virtual/execute/execute.go [hunk 1]
```diff
@@ -222,6 +223,7 @@ func (s *SMExecute) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.
 			panic(throw.IllegalState())
 		}
 	} else if objectState != object.HasState {
+		ctx.Log().Warn("no state on object after readyToWork for object: " + s.execution.Object.String())
 		panic(throw.IllegalState())
```
Comment (bigbes) [inline ledger-core/virtual/execute/execute.go:227]: merge two parts, throw panic with custom message

File: ledger-core/virtual/execute/execute.go [hunk 2]
```diff
@@ -124,6 +124,7 @@ func (s *SMExecute) prepareExecution(ctx context.Context) {
 }
 
 func (s *SMExecute) migrationDefault(ctx smachine.MigrationContext) smachine.StateUpdate {
+	ctx.Log().Trace("migrationDefault")
```
Comment (ivanshibitov) [inline ledger-core/virtual/execute/execute.go:127]: Do you really need? Wouldn't it be written by Conveyor's step logger?

File: ledger-core/virtual/integration/delegated_finished_test.go [hunk 3]
```diff
@@ -195,15 +195,12 @@ func TestVirtual_SendDelegatedFinished_IfPulseChanged_Without_SideEffect(t *test
 	rawWalletState := makeRawWalletState(t, testBalance)
 	msg := makeVStateReportEvent(server.GetPulse().PulseNumber, objectRef, stateID, rawWalletState)
 
-	server.WaitIdleConveyor()
```
Comment (bigbes) [inline ledger-core/virtual/integration/delegated_finished_test.go:198]: sure?

File: ledger-core/virtual/execute/execute.go [hunk 4]
```diff
@@ -222,6 +223,7 @@ func (s *SMExecute) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.
 			panic(throw.IllegalState())
 		}
 	} else if objectState != object.HasState {
+		ctx.Log().Warn("no state on object after readyToWork for object: " + s.execution.Object.String())
```
Comment (ivanshibitov) [inline ledger-core/virtual/execute/execute.go:226]: Warn -> Err may be?

File: ledger-core/virtual/integration/delegated_finished_test.go (hunk 3 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/delegated_finished_test.go:198]: yep, WaitIdleConveyor now in Server Init, and it is ready to serve

File: ledger-core/virtual/integration/method_test.go [hunk 5]
```diff
@@ -358,14 +356,20 @@ func TestVirtual_CallMethodAfterPulseChange(t *testing.T) {
 		server.SendMessage(ctx, msg)
 	}
 
+	server.WaitIdleConveyor()
+	server.ResetActiveConveyorFlag()
```
Comment (bigbes) [inline ledger-core/virtual/integration/method_test.go:360]: No need to reset active conveyor flag, no more

Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: fixed

File: ledger-core/virtual/execute/execute.go (hunk 4 above)
Comment (sexyhamster) [reply ledger-core/virtual/execute/execute.go:225]: fixed

File: ledger-core/virtual/execute/execute.go (hunk 2 above)
Comment (sexyhamster) [reply ledger-core/virtual/execute/execute.go]: removed

File: ledger-core/virtual/execute/execute.go (hunk 1 above)
Comment (sexyhamster) [reply ledger-core/virtual/execute/execute.go]: done

File: ledger-core/virtual/integration/staterequest_test.go [hunk 6]
```diff
@@ -64,14 +64,23 @@ func TestVirtual_VStateRequest_WithoutBody(t *testing.T) {
...
+
+	msg := makeVStateRequestEvent(server.GetPulse().PrevPulseNumber, objectRef, 0)
 	server.SendMessage(ctx, msg)
 
 	select {
 	case data := <-reportChan:
 		assert.Equal(t, &payload.VStateReport{
 			Status:           payload.Ready,
-			AsOf:             server.GetPulse().PulseNumber,
+			AsOf:             server.GetPulse().PrevPulseNumber,
```
Comment (bigbes) [inline ledger-core/virtual/integration/staterequest_test.go:83]: hmm, no, please extract previous pulse from pulsegenerator and use it here

Comment (bigbes) [reply ledger-core/virtual/integration/staterequest_test.go:83]: for consistency with code

File: ledger-core/virtual/virtual.go [hunk 7]
```diff
@@ -68,7 +69,7 @@ func (lr *Dispatcher) Init(ctx context.Context) error {
 		PollingTruncate:   1 * time.Millisecond,
 		SlotPageSize:      1000,
 		ScanCountLimit:    100000,
-		SlotMachineLogger: virtualStateMachine.ConveyorLoggerFactory{},
```
Comment (bigbes) [inline ledger-core/virtual/virtual.go:71]: nonono, pls, add as in tests:
```
	if convlog.UseTextConvLog {
		virtualDispatcher.MachineLogger = convlog.MachineLogger{}
	}
```

File: ledger-core/virtual/object/state_report_test.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+	smObject.SharedState.SetState(HasState)
+
+	smReport := newSMReportWithPulse()
+	smReport.report = smObject.BuildStateReport()
+	smReport.report.ProvidedContent.LatestDirtyState = smObject.BuildLatestDirtyState()
+
+	messageService := messageSenderWrapper.NewServiceMockWrapper(mc)
+	checkMessageFn := func(msg payload.Marshaler) {
+		stateReport, ok := msg.(*payload.VStateReport)
+		require.True(t, ok)
```
Comment (bigbes) [inline ledger-core/virtual/object/state_report_test.go:41]: two times? :)

File: ledger-core/virtual/object/state_report.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+func BuildReportKey(object reference.Global, pulse pulse.Number) ReportKey {
+	return ReportKey{
+		ObjectReference: object,
+		Pulse:           pulse,
+	}
+}
+
+func GetSharedStateReport(ctx smachine.InOrderStepContext, object reference.Global, pn pulse.Number) (SharedReportAccessor, bool) {
```
Comment (bigbes) [inline ledger-core/virtual/object/state_report.go:84]: I suggest to split that into two files:
1. SM
2. Work with shared state

File: ledger-core/virtual/object/state_report.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/assured-ledger/blob/master/LICENSE.md.
+
+package object
```
Comment (bigbes) [inline ledger-core/virtual/object/state_report.go:6]: maybe move it to separate package?

File: ledger-core/virtual/integration/staterequest_test.go (hunk 6 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/staterequest_test.go]: created helper server.GetPrevPulse that return prev pulse

File: ledger-core/virtual/virtual.go (hunk 7 above)
Comment (sexyhamster) [reply ledger-core/virtual/virtual.go]: fixed

File: ledger-core/virtual/object/state_report_test.go (hunk 8 above)
Comment (sexyhamster) [reply ledger-core/virtual/object/state_report_test.go:41]: fixed

File: ledger-core/virtual/execute/execute.go [hunk 11]
```diff
@@ -394,6 +400,7 @@ func (s *SMExecute) stepStartRequestProcessing(ctx smachine.ExecutionContext) sm
 }
 
 func (s *SMExecute) migrateDuringExecution(ctx smachine.MigrationContext) smachine.StateUpdate {
+	ctx.Log().Trace("migrateDuringExecution")
```
Comment (ruz) [inline ledger-core/virtual/execute/execute.go:403]: remove this log

File: ledger-core/virtual/object/state_report.go (hunk 10 above)
Comment (sexyhamster) [reply ledger-core/virtual/object/state_report.go]: done

File: ledger-core/virtual/object/state_report.go (hunk 9 above)
Comment (sexyhamster) [reply ledger-core/virtual/object/state_report.go]: done

File: ledger-core/virtual/execute/execute.go (hunk 11 above)
Comment (sexyhamster) [reply ledger-core/virtual/execute/execute.go]: removed

**Pull request merged**